src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs

```diff
@@ -1299,9 +1299,9 @@ private bool canTailCall(CORINFO_METHOD_STRUCT_* callerHnd, CORINFO_METHOD_STRUC
                     return false;
                 }
 
-                // Do not tailcall from methods that are marked as noinline (people often use no-inline
+                // Do not tailcall from methods that are marked as NoInlining or NoOptimization (people often use no-inline
                 // to mean "I want to always see this method in stacktrace")
-                if (caller.IsNoInlining)
+                if (caller.IsNoInlining || caller.IsNoOptimization)
```

Comment:
Is it necessary to check for NoOptimization here? I would expect that the JIT disables tailcall optimizations when the method is marked NoOptimization.

Reply:
@jkotas um.. jit is not aware that a call has NoOptimization on it so it does emit tail calls for it, [example](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4axAMxNSDITQDeNBqYYmzzJExQMAYhAgAKAJQXTx6me8MAQtihXNQYAehCGDGwpODBsSWkAE3cGAF8aZIBtAFkYDAALCASASXFJJxz8wpKxSQB5MT4ILlwWADkIer58XgAvbEauFwBdZKsbPwDXZM8fMzCGXFEYBmBeAHMGSATlsVgFGB5eLg3eDAYAMyhRFZgjk65JI5gkr1mmcgBOIOTvZi+XDSvWZ/b5AnwggE/SyfUFvd7/QFwiGIt7IqGmNFg34wyFY6EI9Hw2GonEo4GkwmY7xpagpIA===)